Add a quirk for encoded streams support

youennf opened the pull request:
#### 1b6ff0fd60c00812591432e5b8ca48a1e984c8f4
<pre>
Add a quirk for encoded streams support
<a href="https://rdar.apple.com/158736355">rdar://158736355</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=297959">https://bugs.webkit.org/show_bug.cgi?id=297959</a>

Reviewed by Brent Fulgham.

We add createEncodedStreams behind a quirk.
For now, the quirk is not enabled as this is an experiment.

* Source/WebCore/Modules/webauthn/fido/AuthenticatorGetInfoResponse.cpp: fix unified build issue.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/mediastream/RTCEncodedStreamProducer.cpp: Added.
(WebCore::RTCEncodedStreamProducer::create):
(WebCore::RTCEncodedStreamProducer::RTCEncodedStreamProducer):
(WebCore::m_isVideo):
(WebCore::RTCEncodedStreamProducer::initialize):
(WebCore::RTCEncodedStreamProducer::enqueueFrame):
(WebCore::RTCEncodedStreamProducer::writeFrame):
* Source/WebCore/Modules/mediastream/RTCEncodedStreamProducer.h: Added.
* Source/WebCore/Modules/mediastream/RTCEncodedStreams.h: Added.
* Source/WebCore/Modules/mediastream/RTCEncodedStreams.idl: Added.
* Source/WebCore/Modules/mediastream/RTCRtpReceiver.cpp:
(WebCore::RTCRtpReceiver::createEncodedStreams):
* Source/WebCore/Modules/mediastream/RTCRtpReceiver.h:
* Source/WebCore/Modules/mediastream/RTCRtpReceiver.idl:
* Source/WebCore/Modules/mediastream/RTCRtpSender.cpp:
(WebCore::RTCRtpSender::createEncodedStreams):
* Source/WebCore/Modules/mediastream/RTCRtpSender.h:
* Source/WebCore/Modules/mediastream/RTCRtpSender.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldEnableRTCEncodedStreamsQuirk const):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: <a href="https://commits.webkit.org/299280@main">https://commits.webkit.org/299280@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7192ed3400b8c21d9f46f0babb75860e7f0afc3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118490 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38171 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28822 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124662 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70549 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ddf1c751-8e26-4284-b20f-a8a098185372) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120368 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38867 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46753 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/89934 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b4291906-2594-4ed1-b043-8e0c6efbfb79) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121443 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/30967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106242 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70438 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d0edfd24-26d3-4eb3-b848-b879f3a3a11b) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/30026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24353 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68321 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/100405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24542 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/127728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45397 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/34250 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/127728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45761 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102461 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/127728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25005 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/43798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/21789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45267 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50945 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/44730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/48077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46417 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/Modules/mediastream/RTCEncodedStreamProducer.h:
```diff
@@ -0,0 +1,74 @@
+#include "RTCDtlsTransport.h"
+#include "RTCEncodedStreams.h"
+#include "RTCRtpSenderBackend.h"
+#include "RTCRtpTransceiverDirection.h"
```
Are all of these includes necessary? It seems like several of them could be forward declared.

youennf replied:
Right, some or indeed not needed, I will fix this.

brentfulgham commented on Source/WebCore/Modules/mediastream/RTCEncodedStreams.h:
```diff
@@ -0,0 +1,43 @@
+#if ENABLE(WEB_RTC)
+
+#include "ReadableStream.h"
+#include "WritableStream.h"
```
Couldn't these be forward declarations?

brentfulgham commented on Source/WebCore/Modules/mediastream/RTCRtpSender.cpp:
```diff
@@ -34,6 +34,7 @@
 #if ENABLE(WEB_RTC)
 
 #include "ContextDestructionObserverInlines.h"
+#include "Document.h"
```
I wonder if you could avoid including this heavyweight header, since you don't actually interact with `Document`, it's just an object reference you pass to another function, so the compiler shouldn't need this include.

youennf replied:
Document is given to a method that takes a ScriptExecutionContext.
I guess I can change from now the sender method to take a ScriptExecutionContext.

brentfulgham commented on Source/WebCore/Modules/webauthn/fido/AuthenticatorGetInfoResponse.cpp:
```diff
@@ -134,7 +134,9 @@ Vector<uint8_t> encodeAsCBOR(const AuthenticatorGetInfoResponse& response)
         auto transports = *response.transports();
-        deviceInfoMap.emplace(CBORValue(kCtapAuthenticatorGetInfoTransportsKey), toArrayValue(transports.map(WebCore::toString)));
+        deviceInfoMap.emplace(CBORValue(kCtapAuthenticatorGetInfoTransportsKey), toArrayValue(transports.map([](auto transport) {
+            return WebCore::toString(transport);
```
Is this an intended part of your change? Since very WebAuthn-related, not WebRTC related! :-)

youennf replied:
It is a unified build fix.

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -922,6 +922,13 @@ bool Quirks::shouldEnableEnumerateDeviceQuirk() const
+#if ENABLE(WEB_RTC)
+bool Quirks::shouldEnableRTCEncodedStreamsQuirk() const
+{
+    return needsQuirks() && m_quirksData.shouldEnableRTCEncodedStreamsQuirk;
```
It seems like you are missing a line in `void handleFacebookQuirks(...)` to activate the Quirk!

youennf replied:
Yes, I will do this in a follow up.

brentfulgham reviewed (approved):
You seem to be missing the step where you activate the quirk for Facebook, but perhaps you intend that for a future patch.

I think you might be able to avoid some expensive header includes with more forward-declaring, but I think this patch is great as-is.

youennf committed:
Add a quirk for encoded streams support
rdar://158736355
https://bugs.webkit.org/show_bug.cgi?id=297959

Reviewed by Brent Fulgham.

We add createEncodedStreams behind a quirk.
For now, the quirk is not enabled as this is an experiment.

* Source/WebCore/Modules/webauthn/fido/AuthenticatorGetInfoResponse.cpp: fix unified build issue.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/mediastream/RTCEncodedStreamProducer.cpp: Added.
(WebCore::RTCEncodedStreamProducer::create):
(WebCore::RTCEncodedStreamProducer::RTCEncodedStreamProducer):
(WebCore::m_isVideo):
(WebCore::RTCEncodedStreamProducer::initialize):
(WebCore::RTCEncodedStreamProducer::enqueueFrame):
(WebCore::RTCEncodedStreamProducer::writeFrame):
* Source/WebCore/Modules/mediastream/RTCEncodedStreamProducer.h: Added.
* Source/WebCore/Modules/mediastream/RTCEncodedStreams.h: Added.
* Source/WebCore/Modules/mediastream/RTCEncodedStreams.idl: Added.
* Source/WebCore/Modules/mediastream/RTCRtpReceiver.cpp:
(WebCore::RTCRtpReceiver::createEncodedStreams):
* Source/WebCore/Modules/mediastream/RTCRtpReceiver.h:
* Source/WebCore/Modules/mediastream/RTCRtpReceiver.idl:
* Source/WebCore/Modules/mediastream/RTCRtpSender.cpp:
(WebCore::RTCRtpSender::createEncodedStreams):
* Source/WebCore/Modules/mediastream/RTCRtpSender.h:
* Source/WebCore/Modules/mediastream/RTCRtpSender.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldEnableRTCEncodedStreamsQuirk const):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: https://commits.webkit.org/299280@main
